Add Rointe integration

tggm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
This integration provides support for electrical radiators manufactured by Rointe. This is an initial implementation that supports some of Rointe's products: 

- D-Series Radiator
- Towel Rails
- Thermostat.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
The integration uses cloud polling via a REST API  owned by Rointe. Config flow requires the user to login to the API and then select which particular installation they want to add to HA.

It's currently available via HACS on https://github.com/tggm/rointe-hacs/

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/25118

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

tggm committed:
Rointe Integration

tggm commented:
Currently trying to run tests locally as there's a problem with the pyparsing library

tggm commented:
Test run, some failures in other integrations

```
===================================================================================================================================================================== short test summary info ======================================================================================================================================================================
FAILED tests/components/google/test_calendar.py::test_http_api_event - assert False
FAILED tests/components/google/test_calendar.py::test_http_api_all_day_event - assert False
FAILED tests/components/history_stats/test_sensor.py::test_end_time_with_microseconds_zeroed[Europe/Berlin] - AssertionError: assert '0.0' == '12.0'
FAILED tests/components/history_stats/test_sensor.py::test_end_time_with_microseconds_zeroed[America/Chicago] - AssertionError: assert '0.0' == '12.0'
FAILED tests/components/history_stats/test_sensor.py::test_end_time_with_microseconds_zeroed[US/Hawaii] - AssertionError: assert '0.0' == '12.0'
FAILED tests/components/mobile_app/test_webhook.py::test_webhook_handle_get_config - AssertionError: assert {'components'...32.87336, ...} == {'components'...32.87336, ...}
FAILED tests/components/nest/test_camera_sdm.py::test_camera_web_rtc[yaml-config-only] - RuntimeError: coroutine raised StopIteration
FAILED tests/components/nest/test_camera_sdm.py::test_camera_web_rtc[app-creds] - RuntimeError: coroutine raised StopIteration
FAILED tests/components/nest/test_media_source.py::test_camera_image_resize[yaml-config-only] - AssertionError: Response not matched: <ClientResponse(http://127.0.0.1:38129/api/nest/event_media/91a2f758373de5887b1c24a4ec437046/WyJDalk1WTNWS2FUWndSM280WTE5WWJUVmZNRiIsICIxYVhFdmk5YWpLVlRkRHNYZEpkYThmemZDYSJd/thumbnail) [500 Internal Server Error]>
FAILED tests/components/nest/test_media_source.py::test_camera_image_resize[app-creds] - AssertionError: Response not matched: <ClientResponse(http://127.0.0.1:43571/api/nest/event_media/4ecd5784b1cbeb4c7ee9d16fdfeb42b4/WyJDalk1WTNWS2FUWndSM280WTE5WWJUVmZNRiIsICIxYVhFdmk5YWpLVlRkRHNYZEpkYThmemZDYSJd/thumbnail) [500 Internal Server Error]>
FAILED tests/components/roku/test_media_player.py::test_services_play_media_audio[url-http://localhost/media.m4a-media.m4a-m4a] - AssertionError: Expected 'play_on_roku' to be called once. Called 0 times.
FAILED tests/components/roku/test_media_player.py::test_services_play_media_audio[music-http://localhost/media.m4a-media.m4a-m4a] - AssertionError: Expected 'play_on_roku' to be called once. Called 0 times.
FAILED tests/helpers/test_translation.py::test_get_translations - KeyError: 'component.switch.state.string1

Results (40656.76s):
    23574 passed
      13 failed
         - tests/components/google/test_calendar.py:479 test_http_api_event
         - tests/components/google/test_calendar.py:504 test_http_api_all_day_event
         - tests/components/history_stats/test_sensor.py:1387 test_end_time_with_microseconds_zeroed[Europe/Berlin]
         - tests/components/history_stats/test_sensor.py:1387 test_end_time_with_microseconds_zeroed[America/Chicago]
         - tests/components/history_stats/test_sensor.py:1387 test_end_time_with_microseconds_zeroed[US/Hawaii]
         - tests/components/mobile_app/test_webhook.py:256 test_webhook_handle_get_config
         - tests/components/nest/test_camera_sdm.py:572 test_camera_web_rtc[yaml-config-only]
         - tests/components/nest/test_camera_sdm.py:572 test_camera_web_rtc[app-creds]
         - tests/components/nest/test_media_source.py:1336 test_camera_image_resize[yaml-config-only]
         - tests/components/nest/test_media_source.py:1336 test_camera_image_resize[app-creds]
         - tests/components/roku/test_media_player.py:511 test_services_play_media_audio[url-http://localhost/media.m4a-media.m4a-m4a]
         - tests/components/roku/test_media_player.py:511 test_services_play_media_audio[music-http://localhost/media.m4a-media.m4a-m4a]
         - tests/helpers/test_translation.py:94 test_get_translations
      34 xfailed
      13 skipped
```

tggm commented:
Added config_flows tests

bachya commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,133 @@
+                CONF_INSTALLATION: user_input[CONF_INSTALLATION],
+                CONF_USERNAME: self.step_user_data[CONF_USERNAME],
+                CONF_PASSWORD: self.step_user_data[CONF_PASSWORD],
+                CONF_LOCAL_ID: self.step_user_local_id,
```
Is `local_id` the unique characteristic between config entries? If so, we should check that one doesn't already exist (so the user doesn't create duplicates).

tggm replied:
`local_id` is unique for each user account. It's akin to a user_id for the API

bachya replied:
Got it. In that case, we should ensure that case is handled. Example: https://github.com/home-assistant/core/blob/111a38589e56fdd1caf09a020a2ee435da19e946/homeassistant/components/ambient_station/config_flow.py#L40-L41

tggm replied:
Thank you for this. It was indeed allowing the integration to be added multiple times.

bachya commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,133 @@
+            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+        )
+
+    async def async_step_installation(self, user_input=None) -> FlowResult:
```
Be consistent with typing on `user_input` (like you have elsewhere).

bachya commented on homeassistant/components/rointe/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+        )
+
+        for platform in PLATFORMS:
+            self.unregistered_keys[platform] = {}
```
This can be simplified:

```python
self.unregistered_keys = {platform: {} for platform in PLATFORMS}
```

bachya commented on homeassistant/components/rointe/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+        for device_id, device in devices.items():
+            for platform in PLATFORMS:
+                if device_id not in self.unregistered_keys[platform]:
+                    self.unregistered_keys[platform][device_id] = device
```
Nested loops indicate a good spot for a dict comprehension. I can't tell your logic, so I can't give you a specific example.

tggm replied:
Refactored this block and split the call to `device_update_info` outside the loops. The purpose of that call is to update the registry in case a device's firmware is updated (Otherwise it will keep showing as "update available")

bachya commented on homeassistant/components/rointe/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+                for device_id, device in discovered_devices.items():
+                    if device_id in self.unregistered_keys[platform]:
+                        for constructor in entity_constructor_list:
+                            new_entities.append(constructor(device, self))
```
Nested loops indicate a good spot for a dict comprehension. I can't tell your logic, so I can't give you a specific example.

bachya commented on homeassistant/components/rointe/rointe_entity.py:
```diff
@@ -0,0 +1,105 @@
+class RointeHAEntity(CoordinatorEntity):
+    """Rointe entity base class."""
+
+    def __init__(self, coordinator: RointeDataUpdateCoordinator, name, unique_id):
```
Argument typing (since you do this elsewhere)?

bachya commented on homeassistant/components/rointe/rointe_entity.py:
```diff
@@ -0,0 +1,105 @@
+    @property
+    def unique_id(self):
+        """Return the unique id."""
+        return self._unique_id
```
If you store the unique ID in `_attr_unique_id` instead of `_unique_id`, you can avoid redefining this property.

tggm commented on homeassistant/components/rointe/const.py:
```diff
@@ -0,0 +1,31 @@
+DOMAIN = "rointe"
+DEVICE_DOMAIN = "climate"
+ATTRIBUTION = "Data provided by rointeconnect.com"
+PLATFORMS: list[str] = ["climate", "sensor", "update"]
```
Understood. Thank you!

tggm commented on homeassistant/components/rointe/__init__.py:
```diff
@@ -0,0 +1,111 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Rointe Heaters from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
```
Refactored this part

tggm commented on homeassistant/components/rointe/__init__.py:
```diff
@@ -0,0 +1,111 @@
+
+        while coordinator.cleanup_callbacks:
+            coordinator.cleanup_callbacks.pop()()
+
```
I'm not quite sure what to do here. I remember struggling a lot with the coordinator and I ended up merging in my head several bits of code I got from Discord and from other integrations.

I'm guessing the coordinator should also be removed from hass.data? Is this what's intended?

tggm replied:
Refactored this (based on the Fronius integration) and I'm now removing the entire Domain from Hass data.

tggm commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,133 @@
+
+                return await self.async_step_installation(None)
+
+            except Exception as ex:  # pylint: disable=broad-except
```
Refactored this block to check for errors in each API call and return immediately with a `async_show_form` containing a relevant errors dict.

bachya commented:
> This latest commit contains a reauth flow I've been working. It also contains a bunch of debug statements which I added to help another user setup this integration (it's currently available via HACS so I have a couple of people beta testing this).

Per [my previous comment](#discussion_r1041392490), I would recommend that you do a reauth flow in a separate PR—this PR already has plenty in it.

tggm commented:
> Per [my previous comment](#discussion_r1041392490), I would recommend that you do a reauth flow in a separate PR—this PR already has plenty in it.

Will do.

tggm commented:
@bachya I fixed the problem with unloading the integration. It was a stupid error where I was using a class variable to store my entities.

I also updated the code to use `async_forward_entry_setups()` according to [this post](https://developers.home-assistant.io/blog/2022/07/08/config_entry_forwards/)

Finally I removed some useless code from `async_unload_entry` that I'd copied from another integration.

bachya commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+            return self.async_show_form(
+                step_id="user",
+                data_schema=STEP_USER_DATA_SCHEMA,
+                errors={"base": login_error_code.error_message},
```
Reminder that every error message will need to be reflected in `strings.json` for translations—I don't currently see those. If you don't know every possible message (or don't want to manage so many strings), you might consider a generalized message for the UI and logging an error with the actual message.

bachya commented on homeassistant/components/rointe/strings.json:
```diff
@@ -0,0 +1,21 @@
+      }
+    },
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
```
I don't see this string used anywhere in the config flow.

bachya commented on homeassistant/components/rointe/strings.json:
```diff
@@ -0,0 +1,21 @@
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
+      "unknown": "[%key:common::config_flow::error::unknown%]",
```
I don't see this string used anywhere in the config flow.

bachya commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+
+        local_id = local_id_response.data
+
+        installations_response = await self.hass.async_add_executor_job(
```
Does the local ID check need to happen before this call? If not, you could batch those two calls together (and handle errors once).

If order matters, disregard.

tggm replied:
I noticed that I could get that local_id during login thus avoiding one call. Good catch.

bachya commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+                        )
+                    }
+                ),
+                errors=errors,
```
How would we have errors at this stage?

bachya commented on homeassistant/components/rointe/const.py:
```diff
@@ -0,0 +1,34 @@
+
+DOMAIN = "rointe"
+DEVICE_DOMAIN = "climate"
+PLATFORMS: list[str] = ["climate", "sensor", "update"]
```
Instead of strings, use `Platform`—example: https://github.com/home-assistant/core/blob/96c442349f10e7d9b4823d684c127a3ef5a41d0e/homeassistant/components/notion/__init__.py#L31

bachya commented on homeassistant/components/rointe/coordinator.py:
```diff
@@ -0,0 +1,109 @@
+            hass,
+            LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=ROINTE_API_REFRESH_SECONDS),
```
Since this is only used in `coordinator.py`, why not store a constant here?

```python
ROINTE_API_REFRESH_INTERVAL = timedelta(seconds=60)
```

bachya commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+    ) -> FlowResult:
+        """Step: User credentials validation."""
+
+        errors: dict[str, str] = {}
```
Not point in defining these before we check whether the user has provided input.

bachya commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
```
How would we have errors here?

bachya commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_USERNAME): str,
```
We generally use `cv.string` instead of `str` (except in very unique circumstances): https://github.com/home-assistant/core/blob/96c442349f10e7d9b4823d684c127a3ef5a41d0e/homeassistant/helpers/config_validation.py#L558-L569

bachya commented on homeassistant/components/rointe/coordinator.py:
```diff
@@ -0,0 +1,109 @@
+        devices = await self.device_manager.update()
+
+        for platform in PLATFORMS:
+            self.unregistered_keys[platform].update(
```
Is this replacing an entire dict? Or leaving some of the old dict in place?

bachya commented on homeassistant/components/rointe/__init__.py:
```diff
@@ -0,0 +1,74 @@
+    return unload_ok
+
+
+async def init_device_manager(
```
Totally up to you, but since this method is now pretty tight/focused, does it make sense to keep it as a separate method (vs. just putting the contents in `async_setup_entry`)?

No pressure there; just thinking.

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,265 @@
+        self.entity_description = ClimateEntityDescription(
+            key="radiator",
+            name=radiator.name,
+            entity_category=EntityCategory.CONFIG,
```
Isn't this an entity to adjust HVAC? Is there a reason it's a config entity?

bachya commented on homeassistant/components/rointe/device_manager.py:
```diff
@@ -0,0 +1,384 @@
+                LOGGER.debug("No devices info found for zone")
+                continue
+
+            devices = zone["devices"]
```
This check can be simplified:

```python
if (devices := zone.get("devices")) is None:
    LOGGER.debug("No devices info found for zone")
    continue
```

After this point, `devices` will contain what you need?

bachya commented on homeassistant/components/rointe/device_manager.py:
```diff
@@ -0,0 +1,384 @@
+    version = device_data["data"].get("product_version", None)
+    current_firmware = device_data["firmware"].get("firmware_version_device", None)
+
+    if not product_type or not version or not current_firmware:
```
```suggestion
    if not any((product_type, version, current_firmware)):
```

bachya commented on homeassistant/components/rointe/device_manager.py:
```diff
@@ -0,0 +1,384 @@
+
+    product = get_product_by_type_version(product_type, version)
+
+    if not product:
```
This can be simplified:

```python
if (product := get_product_by_type_version(product_type, version)) is None:
```

bachya commented on homeassistant/components/rointe/device_manager.py:
```diff
@@ -0,0 +1,384 @@
+)
+
+
+def determine_latest_firmware(device_data, fw_map) -> str | None:
```
Can this arguments be typed (since you do a great job of typing elsewhere)?

bachya commented on homeassistant/components/rointe/device_manager.py:
```diff
@@ -0,0 +1,384 @@
+
+                if device_data_response.success:
+                    # Retrieve energy stats.
+                    energy_stats_response: ApiResponse = (
```
We're looping through devices and making a lot of synchronous executor jobs—can they be collected and batched?

You might create a separate method that gets device data and, if successful, gets energy stats data—then, batch _that_ method together.

tggm replied:
I don't quite understand what you mean by batching the methods.

Are you referring to iterating in parallel (akin to `joblib.Parallel` )?

Or are you referring to making a single call to get all the device's data and then another call to get all device's energy data? Unfortunately the API doesn't allow me to query all devices in a single call. It's possible with energy data but I have to request a node so high up on their object hierarchy that the response would be massive and would take > 300s to return.

Or something else? :)

bachya replied:
I'm referring to the former more than the latter.

`hass.async_add_executor_job` creates an asynchronous task (really, an asynchronous wrapper and a thread inside which synchronous code is run). You can run multiple asynchronous tasks simultaneously via the `asyncio.gather` call.

As a (overly simple) example, imagine this:

```python
def test() -> None:
    """Run."""
    print("Running the test function")

tasks = []
for i in range(0, 5):
    tasks.append(self.hass.async_add_executor_job(test))
 
await asyncio.gather(*tasks)
```

This will run five iterations of the `test` function in parallel rather than one-by-one. This silly example won't make much difference, but in your world (where API calls are performing I/O), it can make a big difference.

Of course, this assumes that each task _can_ be run in parallel. If you have API calls that _have_ to happen before others, you have to consider whether this can be revised. For example, if we assume that a different function—`test2`—needs to run before `test`, but still want to explore batching things, we might consider:

```python
def test() -> None:
    """Run."""
    print("Running the test function")

def test2() -> None:
    """Run."""
    print("Running the test2 function")

def combined() -> None:
    """Run."""
    test()
    test2()

tasks = []
for i in range(0, 5):
    tasks.append(self.hass.async_add_executor_job(combined))
 
await asyncio.gather(*tasks)
```

tggm replied:
Got it. I think I can make this work!

tggm replied:
Refactored the code and split the logic through a couple of methods. Please tell me if you approve this approach.

Also, do I need to control the degree of parallelism? Does HASS manages that for me?

bachya commented on homeassistant/components/rointe/device_manager.py:
```diff
@@ -0,0 +1,384 @@
+                new_devices.append(rointe_device)
+                self.rointe_devices[device_id] = rointe_device
+
+            except Exception as ex:  # pylint: disable=broad-except
```
Can we have a more specific exception here? It's generally bad form to catch broad exceptions unless it's a "I've caught every other type; I really don't know what this is."

Additionally, it isn't clear which of the elements of this try/except can actually generate the expression. I would suggest more, small try/except blocks to make it very clear.

tggm replied:
Actually there wasn't anything that explicitly throws exceptions and I ended removing the try-except block.

bachya commented on homeassistant/components/rointe/sensor.py:
```diff
@@ -0,0 +1,160 @@
+            native_unit_of_measurement=TEMP_CELSIUS,
+            device_class=SensorDeviceClass.TEMPERATURE,
+            state_class=SensorStateClass.MEASUREMENT,
+            entity_category=EntityCategory.DIAGNOSTIC,
```
Is this the right category? This sensor's only purpose is diagnostics info? Usually, current temperature would just be a "normal" entity.

bachya commented on homeassistant/components/rointe/sensor.py:
```diff
@@ -0,0 +1,160 @@
+            unique_id=f"{radiator.id}-{description.key}",
+        )
+
+        self.entity_description = description
```
If you just set `self.entity_description` when you're creating it, you can eliminate this line?

bachya commented on homeassistant/components/rointe/sensor.py:
```diff
@@ -0,0 +1,160 @@
+    ) -> None:
+        """Init the sensor entity."""
+
+        description = SensorEntityDescription(
```
Do you anticipate this being the only sensor that's ever added?

If not, you might want to plan for easier expansion—check out Notion for an example: https://github.com/home-assistant/core/blob/dev/homeassistant/components/notion/sensor.py

tggm replied:
See below.

bachya commented on homeassistant/components/rointe/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        return None
+
+
+class RointePowerSensor(RointeRadiatorEntity, SensorEntity):
```
These three sensors classes are very similar—I'm wondering if they could be combined into one (with the entity description providing what's necessary to display differences). Check out what I'm proposing for PurpleAir: https://github.com/home-assistant/core/blob/da72a38e383b4fb9071b8e9f1c1f61313fef0231/homeassistant/components/purpleair/sensor.py

tggm replied:
So, I stole your code from Purple Air :blush: which made the whole sensor platform much more elegant.

However I discovered a couple of snags: First I had to move the sensor description list to a different module, to avoid a circular import.

Secondly, I had to add a different method on the Coordinator to deal with the sensors. Basically my code isn't as sophisticated as PurpleAir's or Notion's.

I need to pass a list of Entity classes to the coordinator that then creates the new instance and populates the RointeEntity instance which is where all the data from the radiator is stored. 

On all the platforms used by this integration every entity contains a reference to `self._radiator` which is passed to the new entity on `add_entities_for_seen_keys` and now also on `add_sensor_entities_for_seen_keys`.

Unfortunately I don't have a proper mental model of how `ConfigEntry`, `async_setup_entry`, the DataCoordinator and `async_add_entities()` all should work together to proper refactor this. The current approach was (somehow painfully) put together by looking at other integrations and some comments I got on discord.

I leave to you to determine if you prefer this approach with a new method on the coordinator, the old approach, or something else. :) 

Additional optimizations:

- The `coordinator.add_entities_for_seen_keys()` method had a callback inside it which wasn't really necessary neither called from anywhere else. The methods was simplified.
- I noticed that only the coordinator was the only thing being stored on hass.data, so I removed the dictionary for a direct reference to the coordinator

bachya replied:
Got it. Let's not worry about it for now—there's nothing _wrong_ with these current classes other than some redundancy. If that will unstick you for now, that's fine; you can always return later if a new idea strikes you.

tggm replied:
If you approve I'll leave the current approach with an extra method to parse that sensor descriptions list. It's likely that new sensors will be added anyway.

bachya reviewed (commented):
You're doing great—keep going!

tggm commented:
Hi @bachya,

Are there any other tasks you want me to work to help with the review?

bachya commented on homeassistant/components/rointe/__init__.py:
```diff
@@ -0,0 +1,87 @@
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
```
This can be simplified:

```python
if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
    hass.data[DOMAIN].pop(entry.entry_id)
```

bachya commented on homeassistant/components/rointe/__init__.py:
```diff
@@ -0,0 +1,87 @@
+    return unload_ok
+
+
+async def init_device_manager(
```
Is this method used anymore (now that most of its contents are in `async_setup_entry`)?

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,261 @@
+    )
+
+
+class RointeHaClimate(RointeRadiatorEntity, ClimateEntity, ABC):
```
`ClimateEntity` inherits from `Entity`, which inherits from `ABC`—is there a reason to inherit from `ABC` again here?

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,261 @@
+        """Signal a radiator change."""
+
+        # Update the data
+        await self.coordinator.async_request_refresh()
```
`async_request_refresh` uses the coordinator's debouncer, which will wait about 10 seconds to before updating (to see if multiple calls can be batched). Is that what you want? (totally fine if so—just checking.)

tggm replied:
I wasn't aware of that delay. Honestly I'd prefer that the refresh request is dispatched immediately. (If possible)

The objective here is to request the device's status (from the API) after sending a request to modify something.

The process of modifying a parameter in this integration is:

**Note:** _On the text below "device" means the physical hardware being controlled. not the HA object._

1. The user changes something on the UI (example `climate.async_set_temperature()`)
2. Send the request to the API via the device_manager.
3. Assume the request will be successful (optimistic approach).
4. Update the underlying data object that backs the entities with the new value (for example set the new temperature and other props like hvac_mode)
  i. This will update the UI with the new value. (more on this at the end)
5. Submit the request refresh. (`coordinator.async_request_refresh()`) to load the current state of the device.
   1. In case anything goes wrong on the API the entity's state will be updated to reflect the true state of the device.

Essentially this is to defend against a scenario where a successful call to API (HTTP 200) is rejected by the device for some reason and does not change its state.

If you prefer a different approach I'll be happy to change the code.

Additional question: How could I force the entities to refresh? For example, on my climate platform I have:
```

    @property
    def current_temperature(self) -> float:
        """Get current temperature (Probe)."""
        return self._radiator.temp_probe
```

When the user sets a new temperature I need set other properties like hvac_mode. I notice that it takes a few seconds for those changes to be reflected on the UI. How could I ensure the UI is updated, considering I updated the underlying object that backs all of the platform's properties ?

bachya replied:
> I wasn't aware of that delay. Honestly I'd prefer that the refresh request is dispatched immediately. (If possible)

In that case, you can use `async_refresh`.

> Additional question: How could I force the entities to refresh?

The common way to do this when using a `DataUpdateCoordinator` is to have all of your entities inherit from (`CoordinatorEntity`)[https://github.com/home-assistant/core/blob/d423cbf8ebc3fec1303055b79e7006cc0701c94a/homeassistant/helpers/update_coordinator.py#L349]. By default, these entities subscribe to coordinator updates and update whenever new data is received: https://github.com/home-assistant/core/blob/d423cbf8ebc3fec1303055b79e7006cc0701c94a/homeassistant/helpers/update_coordinator.py#L369-L376

tggm replied:
Thanks! I've changed to `async_refresh`. My entities already inherit from `CoordinatorEntity`.

bachya commented on homeassistant/components/rointe/const.py:
```diff
@@ -0,0 +1,39 @@
+
+DOMAIN = "rointe"
+DEVICE_DOMAIN = "climate"
+PLATFORMS: list[str] = [Platform.CLIMATE, Platform.SENSOR, Platform.UPDATE]
```
If you're going to type this, it would be more accurate to say `list[Platform]`.

bachya commented on homeassistant/components/rointe/sensor_descriptions.py:
```diff
@@ -0,0 +1,74 @@
+    """Define an object to describe Rointe sensor entities."""
+
+
+SENSOR_DESCRIPTIONS = [
```
It's common practice to put these descriptions in the same file containing entity class definitions (in this case, `sensor.py`).

bachya commented on tests/components/rointe/test_config_flow.py:
```diff
@@ -0,0 +1,126 @@
+    with patch(
+        "homeassistant.components.rointe.config_flow.RointeAPI.initialize_authentication",
+        return_value=ApiResponse(True, None, None),
+    ), patch(
```
If you use identical patches between tests, you might consider using a fixture on your integration's `conftest.py`. Example: https://github.com/home-assistant/core/blob/93dd971b399db9f5b1b009e88c57e5b44597432c/tests/components/purpleair/conftest.py#L84-L96

Then, you can include that fixture in your tests. Example: https://github.com/home-assistant/core/blob/93dd971b399db9f5b1b009e88c57e5b44597432c/tests/components/purpleair/test_config_flow.py#L12

tggm commented:
What are the test coverage targets for files other than `config_flow.py` ?

bachya commented:
> What are the test coverage targets for files other than `config_flow.py` ?

Depends on your objective. Config flow tests are required; (almost) everything else is discretionary.

More info: https://developers.home-assistant.io/docs/integration_quality_scale_index?_highlight=coverage#gold-

tggm commented:
> > What are the test coverage targets for files other than `config_flow.py` ?
> 
> Depends on your objective. Config flow tests are required; (almost) everything else is discretionary.
> 
> More info: https://developers.home-assistant.io/docs/integration_quality_scale_index?_highlight=coverage#gold-

Thanks. Right now the objective is getting the integration merged (no-score) and then aim for Silver and up.

I'll keep adding tests in the meantime (assuming that adding new test files is OK during the review process).

I'll remove entries from `.coveragerc` as soon as individual files reach 100%

tggm commented:
@bachya Added a couple more tests and fixed minor issues found during tests

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,272 @@
+    @property
+    def hvac_modes(self) -> list[str]:
+        """Return hvac modes available."""
+        return [HVACMode.OFF, HVACMode.HEAT, HVACMode.AUTO]
```
I would make this a constant (so that it isn't run fresh every time `hvac_modes` is accessed).

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,272 @@
+    @property
+    def preset_modes(self) -> list[str]:
+        """Return the available preset modes."""
+        return [PRESET_COMFORT, PRESET_ECO, PRESET_ROINTE_ICE]
```
I would make this a constant (so that it isn't run fresh every time `preset_modes` is accessed).

bachya commented on tests/components/rointe/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+            {CONF_USERNAME: USERNAME, CONF_PASSWORD: PASSWORD},
+        )
+
+    assert result2["type"] == data_entry_flow.FlowResultType.FORM
```
Tests should either end with `FlowResultType.ABORT` or `FlowResultType.CREATE_FORM`. In cases like this one, test that you can recover from errors.

tggm replied:
I've expanded a test for the case when a duplicated ID is added (which aborts the configuration).

Also, CREATE_FORM doesn't appear to exist. The `FlowResultType` enum only contains the `FORM` value.

bachya replied:
Sorry, that was a mistype: `FlowResultType.CREATE_ENTRY`.

tggm replied:
Got it. Added that missing assert to the happy path test case.

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+    HVACMode,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import TEMP_CELSIUS
```
Per the Actions warning, this constant is deprecated by the `UnitOfTemperature` enum.

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+from .coordinator import RointeDataUpdateCoordinator
+from .rointe_entity import RointeRadiatorEntity
+
+AVAILABLE_HVAC_MODES: list[str] = [HVACMode.OFF, HVACMode.HEAT, HVACMode.AUTO]
```
I would change the typing here to be more explicit.

```suggestion
AVAILABLE_HVAC_MODES: list[HVACMode] = [HVACMode.OFF, HVACMode.HEAT, HVACMode.AUTO]
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+        return RADIATOR_TEMP_STEP
+
+    @property
+    def hvac_modes(self) -> list[str]:
```
```suggestion
    def hvac_modes(self) -> list[HVACMode]:
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+from .rointe_entity import RointeRadiatorEntity
+
+AVAILABLE_HVAC_MODES: list[str] = [HVACMode.OFF, HVACMode.HEAT, HVACMode.AUTO]
+AVAILABLE_PRESETS: list[str] = [PRESET_COMFORT, PRESET_ECO, PRESET_ROINTE_ICE]
```
Should these be enums instead of constants (to be consistent with `HVACMode`)? In general, we're moving away from "buckets" of constants to enums (so they are clearer, more explicit in typing, etc.).

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+        return AVAILABLE_PRESETS
+
+    @property
+    def hvac_mode(self) -> str:
```
```suggestion
    def hvac_mode(self) -> HVACMode:
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+        return HVACMode.HEAT
+
+    @property
+    def hvac_action(self) -> str:
```
```suggestion
    def hvac_action(self) -> HVACAction:
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+        """Convert the device's preset to HA preset modes."""
+
+        if self._radiator.preset == RADIATOR_PRESET_ECO:
+            return PRESET_ECO
```
Should these be enums, as well?

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+        if self._radiator.preset == RADIATOR_PRESET_COMFORT:
+            return PRESET_COMFORT
+        if self._radiator.preset == RADIATOR_PRESET_ICE:
+            return PRESET_ROINTE_ICE
```
Instead of all these return statements, I would make a dict (personally, a class variable) that maps radiator presets to presets.

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+        if (
+            target_temperature < RADIATOR_TEMP_MIN
+            or target_temperature > RADIATOR_TEMP_MAX
+        ):
```
This could be simplified:

```python
if not RADIATOR_TEMP_MIN <= target_temperature <= RADIATOR_TEMP_MAX:
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+                f"Invalid set_humidity value (must be in range 7.0, 30.0): {target_temperature}"
+            )
+
+        # Round to the nearest half value.from
```
Is the `.from` a leftover?

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+        if not await self.device_manager.send_command(
+            self._radiator, CMD_SET_TEMP, rounded_temp
+        ):
+            LOGGER.error(
```
Is this necessary since we're also raising `HomeAssistantError`?

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+        if not await self.device_manager.send_command(
+            self._radiator, CMD_SET_HVAC_MODE, hvac_mode
+        ):
+            LOGGER.error(
```
Is this necessary since we're also raising `HomeAssistantError`?

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,275 @@
+        if not await self.device_manager.send_command(
+            self._radiator, CMD_SET_PRESET, preset_mode
+        ):
+            LOGGER.error(
```
Is this necessary since we're also raising `HomeAssistantError`?

bachya commented on homeassistant/components/rointe/sensor.py:
```diff
@@ -0,0 +1,108 @@
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ENERGY_KILO_WATT_HOUR, POWER_WATT, TEMP_CELSIUS
```
Per the Actions warning, these are all deprecated in favor of `UnitOfEnergy`, `UnitOfPower`, and `UnitOfTemperature`, respectively.

bachya commented on homeassistant/components/rointe/sensor.py:
```diff
@@ -0,0 +1,108 @@
+        entity_category=EntityCategory.DIAGNOSTIC,
+        value_fn=lambda radiator: radiator.energy_data.kwh
+        if radiator.energy_data
+        else None,
```
We shouldn't have inline ternary statements if they break like this. You could always make a separate function for this description's `value_fn` kwarg.

tggm replied:
I'm almost sure this was done by `black`. Still, I changed it to a function.

bachya commented on homeassistant/components/rointe/sensor.py:
```diff
@@ -0,0 +1,108 @@
+        else None,
+        last_reset_fn=lambda radiator: radiator.energy_data.start
+        if radiator.energy_data
+        else None,
```
Avoid break.

bachya commented on homeassistant/components/rointe/sensor.py:
```diff
@@ -0,0 +1,108 @@
+        entity_category=EntityCategory.DIAGNOSTIC,
+        value_fn=lambda radiator: radiator.energy_data.effective_power
+        if radiator.energy_data
+        else None,
```
Avoid break.

bachya commented on homeassistant/components/rointe/sensor.py:
```diff
@@ -0,0 +1,108 @@
+            unique_id=f"{radiator.id}-{description.key}",
+        )
+
+        self.entity_description = description
```
If you're assigning `self.entity_description`, you don't need to set `name` manually—it will automatically be pulled from the entity description.

https://github.com/home-assistant/core/blob/a2935654b9237ddc09cf5ea67c8bd967c660ce91/homeassistant/helpers/entity.py#L320-L327

tggm replied:
I changed this a bit, after removing the name all the sensor's entity names were changed to the format "sensor.current_temperature_2, _3, etc..."

I added a lambda to define a name based on the device's name:

`name_fn=lambda radiator: f"{radiator.name} Current Temperature",`
as for the Entity ID, they remain incrementally named.

Let me know if this is OK.

bachya commented on homeassistant/components/rointe/strings.json:
```diff
@@ -0,0 +1,19 @@
+      }
+    },
+    "error": {
+      "invalid_auth": "Invalid authentication",
```
This is already defined in `homeassistant/strings.json`.

```suggestion
      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
```

bachya commented on tests/components/rointe/common.py:
```diff
@@ -0,0 +1,149 @@
+
+MOCK_GET_DEVICE_RESPONSES = {
+    "device-1": ApiResponse(
+        True,
```
We generally store JSON payloads for tests inside a `fixtures` directory—example: https://github.com/home-assistant/core/tree/dev/tests/components/purpleair/fixtures

Here's an example of how it can be used inside tests: https://github.com/home-assistant/core/blob/a2935654b9237ddc09cf5ea67c8bd967c660ce91/tests/components/purpleair/conftest.py#L76-L81

tggm commented:
Hi @bachya !

Do you have any next steps for me?

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,254 @@
+from .rointe_entity import RointeRadiatorEntity
+
+AVAILABLE_HVAC_MODES: list[HVACMode] = [HVACMode.OFF, HVACMode.HEAT, HVACMode.AUTO]
+AVAILABLE_PRESETS: list[str] = [
```
```suggestion
AVAILABLE_PRESETS: list[RointePreset] = [
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,254 @@
+    @property
+    def icon(self) -> str | None:
+        """Icon of the entity."""
+        return "mdi:radiator"
```
If this is always going to be the same, I would use a class attribute:

```python
class RointeHaClimate(RointeRadiatorEntity, ClimateEntity):
    """Climate entity."""

    _attr_icon = "mdi:radiator"
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,254 @@
+    @property
+    def temperature_unit(self) -> str:
+        """Temperature unit."""
+        return UnitOfTemperature.CELSIUS
```
If this is always going to be the same, I would use a class attribute:

```python
class RointeHaClimate(RointeRadiatorEntity, ClimateEntity):
    """Climate entity."""

    _attr_temperature_unit = UnitOfTemperature.CELSIUS
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,254 @@
+        """Flag supported features."""
+        return (
+            ClimateEntityFeature.TARGET_TEMPERATURE | ClimateEntityFeature.PRESET_MODE
+        )
```
If this is always going to be the same, I would use a class attribute:

```python
class RointeHaClimate(RointeRadiatorEntity, ClimateEntity):
    """Climate entity."""

    _attr_supported_features = ClimateEntityFeature.TARGET_TEMPERATURE | ClimateEntityFeature.PRESET_MODE
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,254 @@
+    @property
+    def target_temperature_step(self) -> float | None:
+        """Temperature step."""
+        return RADIATOR_TEMP_STEP
```
If this is always going to be the same, I would use a class attribute:

```python
class RointeHaClimate(RointeRadiatorEntity, ClimateEntity):
    """Climate entity."""

    _attr_target_temperature_step = RADIATOR_TEMP_STEP
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,254 @@
+    @property
+    def hvac_modes(self) -> list[HVACMode]:
+        """Return hvac modes available."""
+        return AVAILABLE_HVAC_MODES
```
If this is always going to be the same, I would use a class attribute:

```python
class RointeHaClimate(RointeRadiatorEntity, ClimateEntity):
    """Climate entity."""

    _attr_hvac_modes = AVAILABLE_HVAC_MODES
```

bachya commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,254 @@
+    @property
+    def preset_modes(self) -> list[str]:
+        """Return the available preset modes."""
+        return AVAILABLE_PRESETS
```
If this is always going to be the same, I would use a class attribute:

```python
class RointeHaClimate(RointeRadiatorEntity, ClimateEntity):
    """Climate entity."""

    _attr_preset_modes = AVAILABLE_PRESETS
```

bachya commented on homeassistant/components/rointe/const.py:
```diff
@@ -0,0 +1,44 @@
+CMD_SET_TEMP = "cmd_set_temp"
+CMD_SET_PRESET = "cmd_set_preset"
+CMD_HVAC_OFF = "cmd_turn_off"
+CMD_SET_HVAC_MODE = "cmd_set_hvac_mode"
```
Since these are options of a defined set, should they be enums (to be consistent elsewhere)?

bachya commented on homeassistant/components/rointe/const.py:
```diff
@@ -0,0 +1,44 @@
+RADIATOR_TEMP_MAX = 30.0
+
+RADIATOR_MODE_AUTO = "auto"
+RADIATOR_MODE_MANUAL = "manual"
```
Since these are options of a defined set, should they be enums (to be consistent elsewhere)?

bachya commented:
I've reviewed what I can and think this is shaping up nicely, but given its size, I will ask for additional review. A reminder that with new integrations, you'll generally have a faster review if the initial contribution contains only one platform (e.g., sensor).

tggm commented:
> I've reviewed what I can and think this is shaping up nicely, but given its size, I will ask for additional review. A reminder that with new integrations, you'll generally have a faster review if the initial contribution contains only one platform (e.g., sensor).

Thank you for the review @bachya. 

I understand that it will take time.

tggm commented:
Hi @bachya,

How long do you think it will take for the second opinion review? It has been 2 months :cry:

bachya commented:
I don't know, unfortunately. I'm currently busy at a new job, so I'm not as focused in HASS for the time being.

I recommend you go to [Discord](https://www.home-assistant.io/join-chat/), go to the `#devs_core` channel, and see if anyone would be willing to help you push this across the finish line.

gjohansson-ST reviewed (changes requested):
Please strip it down to an implementation of a single platform https://developers.home-assistant.io/docs/review-process
Also it has conflicts that needs to be addressed.

tggm commented:
> Please strip it down to an implementation of a single platform https://developers.home-assistant.io/docs/review-process Also it has conflicts that needs to be addressed.

I'm sorry, after all the effort and the time this took back and forth in reviews with other people you're are seriously asking me to strip the code down?!

MartinHjelmare commented:
Yes, please remove all but one platform. Downsizing a PR is the fastest way to get it merged. And now you even have a reviewer that has requested changes, which normally means they will merge the PR when all the requested changes are complete.

Thanks for your understanding!

tggm commented:
> Yes, please remove all but one platform. Downsizing a PR is the fastest way to get it merged. And now you even have a reviewer that has requested changes, which normally means they will merge the PR when all the requested changes are complete.
> 
> Thanks for your understanding!

Removed the other platforms (update and sensor).

gjohansson-ST commented on homeassistant/components/rointe/.gitignore:
```diff
@@ -0,0 +1 @@
+*.code-workspace
```
Remove this file

gjohansson-ST commented on homeassistant/components/rointe/const.py:
```diff
@@ -0,0 +1,52 @@
+
+ROINTE_SUPPORTED_DEVICES = ["radiator", "towel", "therm", "radiatorb"]
+
+RADIATOR_DEFAULT_TEMPERATURE = 20
+
+PRESET_ROINTE_ICE = "ice"
+
+RADIATOR_TEMP_STEP = 0.5
+RADIATOR_TEMP_MIN = 7.0
+RADIATOR_TEMP_MAX = 30.0
+
+
+class RointePreset(StrEnum):
+    """Rointe radiators preset modes."""
+
+    ECO = "eco"
+    COMFORT = "comfort"
+    ICE = "ice"
+    NONE = "none"
+    OFF = "off"
+
+
+class RointeCommand(StrEnum):
+    """Device commands."""
+
+    SET_TEMP = "cmd_set_temp"
+    SET_PRESET = "cmd_set_preset"
+    SET_HVAC_MODE = "cmd_set_hvac_mode"
+
+
+class RointeOperationMode(StrEnum):
+    """Device operation mode."""
+
+    AUTO = "auto"
+    MANUAL = "manual"
```
Single platform constants should be in the platform file directly

tggm replied:
Should or must? what would be gained since after this I'll be opening a new PR to add the sensor and update platforms ?

gjohansson-ST replied:
Anything which is only used in a single platform should stay in that particular platform

tggm replied:
Moved the temperature constants back to climate.py. The other enums are required by the device_manager.py also

gjohansson-ST commented on homeassistant/components/rointe/coordinator.py:
```diff
@@ -0,0 +1,156 @@
+ROINTE_API_REFRESH_INTERVAL = timedelta(seconds=60)
+
+
+@dataclass
+class RointeSensorEntityDescriptionMixin:
+    """Define a description mixin for Rointe sensor entities."""
+
+    last_reset_fn: Callable[[RointeDevice], datetime | None]
+    name_fn: Callable[[RointeDevice], str]
+    value_fn: Callable[[RointeDevice], StateType]
+
+
+@dataclass
+class RointeSensorEntityDescription(
+    SensorEntityDescription, RointeSensorEntityDescriptionMixin
+):
+    """Define an object to describe Rointe sensor entities."""
```
Not used

gjohansson-ST commented on homeassistant/components/rointe/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/rointe",
+  "requirements": ["rointe-sdk==1.3.4"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
```
Remove not used keys

gjohansson-ST commented on homeassistant/components/rointe/rointe_entity.py:
```diff
@@ -0,0 +1,63 @@
+"""Rointe devices entity model."""
```
Can we rename the file to just `entity.py`

gjohansson-ST commented on homeassistant/components/rointe/translations/en.json:
```diff
@@ -0,0 +1,19 @@
+{
```
Remove file

gjohansson-ST reviewed (changes requested):
First pass for cleanups.
Please rebase after the cleanups.

emontnemery commented on homeassistant/components/rointe/entity.py:
```diff
@@ -0,0 +1,63 @@
+from .device_manager import RointeDevice, RointeDeviceManager
+
+
+class RointeHAEntity(CoordinatorEntity):
```
```suggestion
class RointeBaseEntity(CoordinatorEntity):
```

emontnemery commented on homeassistant/components/rointe/entity.py:
```diff
@@ -0,0 +1,63 @@
+    ) -> None:
+        """Initialize the entity."""
+        super().__init__(coordinator)
+        self._attr_unique_id = f"rointe-{unique_id}"
```
Why do we prefix with `rointe`? A unique id does not need to be globally unique, it only needs to be unique within a platform.
This means unique ids must be unique among all rointe climate entities, but a zha climate entity or a rointe sensor can have the same unique id as a rointe climate entity without a collision.

tggm replied:
Is it necessary to remove the prefix?

gjohansson-ST replied:
Why would you not do it as it's not necessary  and holding unnecessary data seems wrong

tggm replied:
There are already some people that use this integration via [HACS](https://github.com/tggm/rointe-hacs). There's also a person testing it together with the [Versatile Thermostat](https://github.com/jmcollin78/versatile_thermostat).

I really appreciate having other people using, testing and debugging this integration which gives me confidence of its stability. Will renaming the entities break their integrations/dashboards? Because I would very much like to avoid that unless this change is conditional to your approval of this PR.

MartinHjelmare replied:
We don't accept unnecessary code to stay compatible with custom integrations. It's not recommended to have custom integrations with the same domain as core integrations anyway.

One of the benefits of having the integrations in core is the code review that makes sure that the code follows our guidelines and reaches a certain quality level.

tggm replied:
done

tggm commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,237 @@
+
+        return RADIATOR_TEMP_MIN
+
+    @property
+    def target_temperature_high(self) -> float:
+        """Max selectable target temperature."""
+        if self._radiator.user_mode_supported and self._radiator.user_mode:
+            return self._radiator.um_max_temp
+
+        return RADIATOR_TEMP_MAX
```
I will remove these two methods since the devices do not support HEAT_COOL

tggm committed:
PR fixes

tggm committed:
PR Fixes

joostlek commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,227 @@
+
+        super().__init__(coordinator, radiator, unique_id=radiator.id)
+
+        self.entity_description = ClimateEntityDescription(
+            key="radiator",
+            name=radiator.name,
+        )
```
This isn't needed. I think you added this to be able to set the name. Instead you can create a DeviceInfo with the radiator name as name, and because you have `_attr_name = None`, the name of the entity will be set to the name of the device.

tggm replied:
The `DeviceInfo` is provided by `RointeRadiatorEntity`.

`class RointeHaClimate(RointeRadiatorEntity, ClimateEntity):`

joostlek commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,227 @@
+    def target_temperature(self) -> float | None:
+        """Return the current temperature or None if the device is off."""
+
+        LOGGER.debug(
+            f"[{self._radiator.name}] :: target_temperature >> Mode: {self._radiator.mode} Power: {self._radiator.power}. Preset: {self._radiator.preset}"
+        )
```
Please don't have lines longer than 88 characters

joostlek commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,227 @@
+        """Convert the device's preset to HA preset modes."""
+
+        # Also captures "none" (man mode, temperature outside presets)
+        return ROINTE_HASS_MAP.get(self._radiator.preset, None)
```
```suggestion
        return ROINTE_HASS_MAP.get(self._radiator.preset)
```

joostlek commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,227 @@
+        LOGGER.debug("Setting temperature to %s", target_temperature)
+
+        if not RADIATOR_TEMP_MIN <= target_temperature <= RADIATOR_TEMP_MAX:
+            raise ValueError(
```
I think you should raise HomeAssistantError here

joostlek commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,227 @@
+
+        # Update the data
+        await self.coordinator.async_refresh()
+        self.async_write_ha_state()
```
I don't think you need to write HA state here, because the moment that the coordinator updates, `_handle_coordinator_callback` is called, which also writes ha state, so this now creates 2 state writes instead of 1

joostlek commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+        self.step_user_data = user_input
+        self.step_user_installations = installations
+
+        return await self.async_step_installation(None)
+
+    async def async_step_installation(
+        self, user_input: dict[str, Any] | None
```
```suggestion
        return await self.async_step_installation()

    async def async_step_installation(
        self, user_input: dict[str, Any] | None = None
```

joostlek commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+
+    async def async_step_installation(
+        self, user_input: dict[str, Any] | None
+    ) -> FlowResult:
```
Okay so apparently 1 account can lead to more installations? Can you maybe elaborate on what an installation is? Usually we have the guideline that 1 (cloud) account = 1 config entry.

tggm replied:
Essentially it represents a house/home. It's part of their hierarchy: Installations -> Zones ->  Devices.

During configuration the user must pick one installation (for the majority of cases there will only be one option available). The Installation ID is then stored as it is required to retrieve devices.

joostlek commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+    def __init__(self) -> None:
+        """Config flow init."""
+        self.step_user_data: dict[str, Any] | None = None
+        self.step_user_local_id: str | None = None
```
unused

joostlek commented on homeassistant/components/rointe/coordinator.py:
```diff
@@ -0,0 +1,105 @@
+            update_interval=ROINTE_API_REFRESH_INTERVAL,
+        )
+
+        self.unregistered_keys = {platform: {} for platform in PLATFORMS}
```
What's a key?

tggm replied:
In that context a "key" is the unique identifier of each device. It's determined by the manufacturer's API

joostlek replied:
By the way, do you have discord? Can you maybe pop me a message? (@joostlek)

joostlek replied:
```suggestion
```

tggm committed:
PR Fixes

joostlek commented on homeassistant/components/rointe/coordinator.py:
```diff
@@ -0,0 +1,105 @@
+
+    async def _async_update_data(self) -> dict[str, RointeDevice]:
+        """Fetch data from API."""
+
+        new_devices = await self.device_manager.update()
+
+        for platform in PLATFORMS:
+            self.unregistered_keys[platform].update(
+                {
+                    device_id: device
+                    for device_id, device in new_devices.items()
+                    if device_id not in self.unregistered_keys[platform]
+                }
+            )
+
+        for device in new_devices.values():
+            device_update_info(self.hass, device)
+
+        return new_devices
```
```suggestion

        return await self.device_manager.update()
```

joostlek commented on homeassistant/components/rointe/coordinator.py:
```diff
@@ -0,0 +1,105 @@
+
+        return new_devices
+
+    @callback
+    def add_entities_for_seen_keys(
+        self,
+        async_add_entities: AddEntitiesCallback,
+        entity_constructor_list: list[Any],
+        platform: str,
+    ) -> None:
+        """Add entities for new devices, for a given platform.
+
+        Called from a platform's `async_setup_entry`.
+        """
+
+        discovered_devices: dict[str, RointeDevice] = self.data
+
+        if not discovered_devices:
+            return
+
+        new_entities: list = []
+
+        for device_id, device in discovered_devices.items():
+            if device_id in self.unregistered_keys[platform]:
+                new_entities.extend(
+                    [
+                        constructor(device, self)
+                        for constructor in entity_constructor_list
+                    ]
+                )
+
+                self.unregistered_keys[platform].pop(device_id)
+
+        if new_entities:
+            async_add_entities(new_entities)
```
```suggestion
```

joostlek commented on homeassistant/components/rointe/coordinator.py:
```diff
@@ -0,0 +1,105 @@
+
+        if new_entities:
+            async_add_entities(new_entities)
+
+
+@callback
+def device_update_info(hass: HomeAssistant, rointe_device: RointeDevice) -> None:
+    """Update device registry info."""
+
+    LOGGER.debug("Updating device registry info for %s", rointe_device.name)
+
+    dev_registry = dr.async_get(hass)
+
+    if device := dev_registry.async_get_device(
+        identifiers={(DOMAIN, rointe_device.id)},
+    ):
+        dev_registry.async_update_device(
+            device.id, sw_version=rointe_device.firmware_version
+        )
```
```suggestion
```

joostlek commented on homeassistant/components/rointe/climate.py:
```diff
@@ -0,0 +1,216 @@
+    """Set up the radiator climate entity from the config entry."""
+    coordinator: RointeDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+
+    # Register the Entity classes and platform on the coordinator.
+    coordinator.add_entities_for_seen_keys(
+        async_add_entities, [RointeHaClimate], "climate"
+    )
```
```suggestion
    async_add_entities(
        RointeHaClimate(device, coordinator)
        for device in coordinator.data.values()
    )
```

tggm commented on homeassistant/components/rointe/config_flow.py:
```diff
@@ -0,0 +1,120 @@
+                data_schema=vol.Schema(
+                    {
+                        vol.Required(CONF_INSTALLATION): vol.In(
+                            self.step_user_installations
```
There will be a list of names yes. Each installation has a name. Example:

![image](https://github.com/home-assistant/core/assets/4171985/cd7eecfb-5dbd-4408-ab42-2b8193ec3d9c)
